src/components/form/InputAmount/index.tsx

```diff
@@ -20,55 +24,83 @@ interface Props {
   splitFrom: SplitFrom
 }
 
+const canSetPositionBalance = (
+  positionsLoading: boolean,
+  balancesLoading: boolean,
+  positions: GetMultiPositions_positions[],
+  balances: BigNumber[],
+  positionIds: string[]
+) => {
+  return (
+    !positionsLoading &&
+    !balancesLoading &&
+    positions.length &&
+    balances.length &&
+    balances.length === positionIds.length &&
+    positions.length === positionIds.length &&
+    JSON.stringify(positions.map(({ id }) => id).sort()) === JSON.stringify([...positionIds].sort())
+  )
+}
+
+const fetchBalance = (
+  provider: Web3Provider | InfuraProvider,
+  signer: JsonRpcSigner,
+  tokenAddress: string,
+  walletAddress: string
+) => {
+  const erc20Service = new ERC20Service(provider, tokenAddress, signer)
+  return erc20Service.balanceOf(walletAddress)
+}
+
 export const InputAmount = ({
   collateral,
   formMethods: { control, setValue },
   positionId,
   splitFrom,
 }: Props) => {
-  const { _type: status, CTService, address, provider, signer } = useWeb3ConnectedOrInfura()
+  const { address, networkConfig, provider, signer } = useWeb3ConnectedOrInfura()
+  const { loading: positionsLoading, positionIds, positions } = useMultiPositionsContext()
+
+  const { balances, loading: balancesLoading } = useBatchBalanceContext()
 
   const [balance, setBalance] = useState<Maybe<BigNumber>>(null)
-  const regexpPosition: RegExp = BYTES_REGEX
+  const [decimals, setDecimals] = useState(0)
 
   useEffect(() => {
     setValue('amount', ZERO_BN)
   }, [collateral, positionId, setValue])
 
   useEffect(() => {
-    let isSubscribed = true
-    if (status === Web3ContextStatus.Connected && address && signer) {
-      const fetchBalance = async () => {
-        if (splitFrom === 'position' && regexpPosition.test(positionId)) {
-          const balance = await CTService.balanceOf(positionId)
-          if (isSubscribed) {
-            setBalance(balance)
-          }
-        } else if (splitFrom === 'collateral' && signer) {
-          const erc20Service = new ERC20Service(provider, collateral.address, signer)
-          const balance = await erc20Service.balanceOf(address)
-          if (isSubscribed) {
-            setBalance(balance)
-          }
-        }
-      }
-
-      fetchBalance()
+    if (
+      splitFrom === 'position' &&
+      canSetPositionBalance(positionsLoading, balancesLoading, positions, balances, positionIds)
+    ) {
+      // FIXME - this only works with non custom tokens
```

Comment:
Please use TODO, the linter throws a warning about it, or create an issue for this